tiled/server/router.py

```diff
@@ -283,141 +90,419 @@ def patch_route_signature(route, query_registry):
-    # we modify the signature here, at runtime, just before handing it to
-    # FastAPI in the usual way.
-
-    # When FastAPI calls the function with these added parameters, they will be
-    # accepted via **filters.
-
-    # Make a copy of the original parameters.
-    signature = inspect.signature(route)
-    parameters = list(signature.parameters.values())
-    # Drop the **filters parameter from the signature.
-    del parameters[-1]
-    # Add a parameter for each field in each type of query.
-    for name, query in query_registry.name_to_query_type.items():
-        for field in dataclasses.fields(query):
-            # The structured "alias" here is based on
-            # https://mglaman.dev/blog/using-json-router-query-your-search-router-indexes
-            if getattr(field.type, "__origin__", None) is list:
-                field_type = str
+    def inner(route: Callable[P, T]) -> Callable[P, T]:
+        # Build a wrapper so that we can modify the signature
+        # without mutating the wrapped original.
+
+        async def route_with_sig(*args, **kwargs):
+            return await route(*args, **kwargs)
+
+        # Black magic here! FastAPI bases its validation and auto-generated swagger
+        # documentation on the signature of the route function. We do not know what
+        # that signature should be at compile-time. We only know it once we have a
+        # chance to check the user-configurable registry of query types. Therefore,
+        # we modify the signature here, at runtime, just before handing it to
+        # FastAPI in the usual way.
+
+        # When FastAPI calls the function with these added parameters, they will be
+        # accepted via **filters.
+
+        # Make a copy of the original parameters.
+        signature = inspect.signature(route)
+        parameters = list(signature.parameters.values())
+        # Drop the **filters parameter from the signature.
+        del parameters[-1]
+        # Add a parameter for each field in each type of query.
+        for name, query in query_registry.name_to_query_type.items():
+            for field in dataclasses.fields(query):
+                # The structured "alias" here is based on
+                # https://mglaman.dev/blog/using-json-router-query-your-search-router-indexes
+                if getattr(field.type, "__origin__", None) is list:
+                    field_type = str
+                else:
+                    field_type = field.type
+                injected_parameter = inspect.Parameter(
+                    name=f"filter___{name}___{field.name}",
+                    kind=inspect.Parameter.POSITIONAL_OR_KEYWORD,
+                    default=Query(None, alias=f"filter[{name}][condition][{field.name}]"),
+                    annotation=Optional[List[field_type]],
+                )
+                parameters.append(injected_parameter)
+        route_with_sig.__signature__ = signature.replace(parameters=parameters)
```

Comment:
I wonder if much of this could be replaced by replacing search/distinct **kwargs with  `Unpack[QueryParams]` where QueryParams is a TypedDict generated by the QueryRegistry?